test/unit/analysis/epacamd_eia_test.py

```diff
@@ -0,0 +1,198 @@
+        "2019-12-31 22:00:00+00:00",
+        "2019-12-31 23:00:00+00:00",
+        "2019-12-31 22:00:00+00:00",
+        "2019-12-31 23:00:00+00:00",
+        "2019-12-31 22:00:00+00:00",
+        "2019-12-31 23:00:00+00:00",
+        "2019-12-31 22:00:00+00:00",
+        "2019-12-31 23:00:00+00:00",
+    ]
+    cems["operating_datetime_utc"] = cems["operating_datetime_utc"].astype(
+        "datetime64[ns, UTC]"
+    )
+    cems["plant_id_eia"] = [10, 10, 10, 10, 10, 10, 11, 11, 11, 11, 12, 12]
+    cems["emissions_unit_id_epa"] = [
+        "a",
+        "a",
+        "b",
+        "b",
+        "c",
+        "c",
+        "a",
+        "a",
+        "b",
+        "b",
+        "a",
+        "a",
+    ]
+    cems["gross_load_mw"] = 0  # not needed for crosswalk testing
+
+    return cems
+
+
+def test__get_unique_keys(mock_cems_extended):
+    """Test that dask and pandas dataframes give the same unique keys."""
+    expected = mock_cems_extended[::2][["plant_id_eia", "emissions_unit_id_epa"]]
+    actual = cw_analysis._get_unique_keys(mock_cems_extended)
+    assert_frame_equal(actual, expected)
+
+    dask_cems = dd.from_pandas(mock_cems_extended, npartitions=2)
+    actual = cw_analysis._get_unique_keys(dask_cems)
+    assert_frame_equal(actual, expected)
+
+
+def test__convert_global_id_to_composite_id(mock_cems_extended):
+    """Test conversion of global_subplant_id to a composite subplant_id.
+
+    The global_subplant_id should be equivalent to the composite (plant_id_eia,
+    subplant_id).
+
+      global_subplant_id  subplant_id  plant_id_eia  emissions_unit_id_epa  generator_id
+    0         0               0               10            a                 0   # one to one
+    1         1               1               10            b                 1   # many to many
+    2         1               1               10            c                 1
+    3         1               1               10            c                 2
+    4         2               0               11            a                 0   # one to many
+    5         2               0               11            b                 0
+    6         3               0               12            a                 0   # many to one
+    7         3               0               12            a                 1
+
+    """
+    uniques = mock_cems_extended[
+        mock_cems_extended["operating_datetime_utc"] == "2019-12-31 22:00:00+00:00"
+    ]
+
+    # simulate join by duplicating rows as appropriate
+    one_to_many = uniques.query('plant_id_eia == 12 and emissions_unit_id_epa == "a"')
+    many_to_many = uniques.query('plant_id_eia == 10 and emissions_unit_id_epa == "c"')
+    expected = (
+        pd.concat([uniques, one_to_many, many_to_many])
+        .sort_index()
+        .reset_index(drop=True)
+    )
+
+    expected = expected.assign(
+        plant_id_eia=expected["plant_id_eia"],
+        emissions_unit_id_epa=expected["emissions_unit_id_epa"],
```

Comment:
Do the first two assignments do anything?

Reply:
Ya, they pull in the plant and emissions ids from another table that was created. It's a little bit of a 'round-about, but I didn't feel like editing the previous version more than I had to.

Reply:
Isn't it just referencing itself to assign a column that already exists?